src/names/name-resolution.md

```diff
@@ -1,4 +1,351 @@
+
+The expansion process is iterative, alternately resolving imports, resolving and expanding macro invocations, then repeating until there are no further macros invocations to resolve. Once this process is completed all the imports are resolved again to ensure that the macro expansion process did not introduce any new ambiguious imports.
+
+> Note
+>
+> This causes so called time traveling ambiguities, such as when a glob import introduces an item that is ambiguous with its own base path.
+>
+> ```rust,compile_fail,E0659
+> macro_rules! m {
+>     () => { mod bar {} }
+> }
+>
+> mod bar {
+>     pub(crate) use m;
+> }
+>
+> fn f() {
+>     // * initially speculatively resolve bar to the module in the crate root
+>     // * expansion of m introduces a second bar module inside the body of f
+>     // * expansion-time resolution finalizes resolutions by re-resolving all
+>     //   imports and macro invocations, sees the introduced ambiguity
+>     //   and reports it as an error
+>     bar::m!(); // ERROR `bar` is ambiguous
+> }
+> ```
+
+r[names.resolution.expansion.imports]
+
+All use declarations are fully resolved during this stage of resolution. Type-relative paths cannot be resolved at this stage of compilation and will produce an error.
+
+```rust,compile_fail,E0432
+mod my_mod {
+    pub const Const: () = ();
+
+    pub enum MyEnum {
+        MyVariant
+    }
+
+    impl MyEnum {
+        pub const Const: () = ();
+    }
+
+    pub type TypeAlias = MyEnum;
+}
+
+fn foo() {
+    // imports resolved at expansion-time
+    use my_mod::MyEnum; // OK
+    use my_mod::MyEnum::MyVariant; // OK
+    use my_mod::TypeAlias; // OK
+    use my_mod::TypeAlias::MyVariant; // Doesn't work
+    use my_mod::MyEnum::Const; // Doesn't work
+    use my_mod::Const; // OK
+
+    // expressions resolved during type-relative resolution
+    let _ = my_mod::TypeAlias::MyVariant; // OK
+    let _ = my_mod::MyEnum::Const; // OK
+}
+```
+
+r[names.resolution.expansion.imports.shadowing]
+
+The following is a list of situations where shadowing of use declarations is permitted:
+
+* [use glob shadowing]
+* [macro textual scope shadowing]
+
+r[names.resolution.expansion.imports.ambiguity]
+## Ambiguities
+
+r[names.resolution.expansion.imports.ambiguity.intro]
+Some situations are an error when there is an ambiguity as to which name a `use` declaration refers. This happens when there are two name candidates that do not resolve to the same entity where neither import is [permitted] to shadow the other.
+
+r[names.resolution.expansion.imports.ambiguity.globvsglob]
+Names may not be resolved through ambiguous glob statements. Glob imports are allowed to import conflicting names in the same namespace as long as the name is not used. Conflicting names from ambiguous glob statements may still be shadowed and used without producing an error. The errors occur at time of use, not time of import.
```

Comment:
```
Conflicting names from ambiguous glob statements may still be ... used
```
Weird wording, they both can and cannot be used.
I guess the intent was to say that the ambiguous globs can be shadowed, and the thing that shadows then can then be used?